src/Adyen/HttpClient/CurlClient.php

```diff
@@ -2,281 +2,282 @@
 
 namespace Adyen\HttpClient;
 
-
 class CurlClient implements ClientInterface
 {
-    /**
-     * Json API request to Adyen
-     *
-     * @param \Adyen\Service $service
-     * @param $requestUrl
-     * @param $params
-     * @return mixed
-     * @throws \Adyen\AdyenException
-     */
-    public function requestJson(\Adyen\Service $service, $requestUrl, $params)
-    {
-        $client = $service->getClient();
-        $config = $client->getConfig();
-        $logger = $client->getLogger();
-        $username = $config->getUsername();
-        $password = $config->getPassword();
-        $xApiKey = $config->getXApiKey();
-
-        $jsonRequest = json_encode($params);
-
-        // log the request
-        $this->logRequest($logger, $requestUrl, $params);
-
-        //Initiate cURL.
-        $ch = curl_init($requestUrl);
-
-        //Tell cURL that we want to send a POST request.
-        curl_setopt($ch, CURLOPT_POST, 1);
-
-        //Attach our encoded JSON string to the POST fields.
-        curl_setopt($ch, CURLOPT_POSTFIELDS, $jsonRequest);
-
-        //create a custom User-Agent
-        $userAgent = $config->get('applicationName') . " " . \Adyen\Client::USER_AGENT_SUFFIX . $client->getLibraryVersion();
-
-        //Set the content type to application/json and use the defined userAgent
-        $headers = array(
-            'Content-Type: application/json',
-            'User-Agent: ' . $userAgent
-        );
-
-        // set authorisation credentials according to support & availability
-        if ($service->requiresApiKey() && !empty($xApiKey)) {
-            //Set the content type to application/json and use the defined userAgent along with the x-api-key
-            $headers[] = 'x-api-key: ' . $xApiKey;
-        } elseif ($service->requiresApiKey() && empty($xApiKey)) {
-            $msg = "Please provide a valid Checkout API Key";
-            throw new \Adyen\AdyenException($msg);
-        } else {
-
-            //Set the basic auth credentials
-            curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
-            curl_setopt($ch, CURLOPT_USERPWD, $username . ":" . $password);
-        }
-
-        //Set the timeout
-        if($config->getTimeout() != null){
-            curl_setopt($ch, CURLOPT_TIMEOUT, $config->getTimeout());
-        }
-
-        //Set the headers
-        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
-
-        // return the result
-        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
-
-        //Execute the request
-        list($result, $httpStatus) = $this->curlRequest($ch);
-
-        // log the raw response
-        $logger->info("JSON Response is: " . $result);
-
-        // result not 200 throw error
-        if ($httpStatus != 200 && $result) {
-            $this->handleResultError($result, $logger);
-        } elseif (!$result) {
-
-            list($errno, $message) = $this->curlError($ch);
-
-            curl_close($ch);
```

Comment:
curlError at L 83 will throw an exception so curl_close won't be called.
Better refactor with this sequence:
```
$response = curl_exec($curl);
$err = curl_error($curl);
curl_close($curl);
```
and then call handleResultError or handleCurlError